Implement the MultiScheduler

ltseeley opened the pull request:
Creates a MultiScheduler that runs multiple sub-schedulers in parallel
and verifies that they all return the same results.

Signed-off-by: Logan Seeley <seeley@bitwise.io>

ltseeley commented on libtransact/src/scheduler/multi/core.rs:
```diff
@@ -0,0 +1,132 @@
+            // Receive batch result callbacks from the sub-schedulers
+            match self.rx.recv() {
+                Ok(MultiSchedulerCoreMessage::BatchResult(scheduler_index, batch_result)) => {
+                    let batch_result = batch_result.expect("Received batch result that is None");
```
I'm not totally sure what a batch_result of `None` means; based on the docstring of the Scheduler trait definition, it means "all batch results have been sent", but that's not very clear.

vaporos replied:
It essentially indicates that you can shutdown this side of the channel and exit the loop.

vaporos replied:
Well, it doesn't mean that in this specific context.

ltseeley replied:
So is this expect fine, or is there a case that we should handle where it could be None?

peterschwarz replied:
I think we should maybe revisit this part of the API, as it's not quite how we discussed the use of the callback.    

Recapping an internal discussion on this, though, the None should be handled correctly, not panic.  There are potential scenarios where this None should occur as expected, under the current API.

ltseeley commented on libtransact/src/scheduler/multi/mod.rs:
```diff
@@ -0,0 +1,138 @@
+    pub fn new(
+        mut schedulers: Vec<Box<dyn Scheduler + Send>>,
+        executor: &Executor,
+    ) -> MultiScheduler {
```
The external interface of the MultiScheduler is different from the other schedulers; it takes the sub-schedulers as an argument and it requires access to the Executor to execute the sub-schedulers. Also, since it is not really an actual scheduler itself, its `take_task_iterator` and `new_notifer` methods aren't necessary, so I have left them unimplemented. I'm not sure there's a way to provide the same external interface as the other schedulers.

ltseeley commented on libtransact/src/scheduler/multi/shared.rs:
```diff
@@ -0,0 +1,89 @@
+        self.schedulers.iter_mut().for_each(|scheduler| {
+            scheduler.add_batch(batch.clone());
+        });
+        self.pending_results.insert(batch, (None, HashSet::new()));
```
How should duplicate batches be handled? If a duplicate batch is added, the pending result for that batch will be overridden here.

vaporos replied:
I think adding duplicate batches to the scheduler should be considered a programming error on behalf of the calling code (the code using the scheduler).

ltseeley replied:
Are you suggesting we should panic here? Or just ignore this possibility?

vaporos replied:
Or possibly add_batch() in the scheduler needs to return a result so we should propagate up the error.

ltseeley replied:
I'm not sure returning a result from add_batch is particularly interesting for schedulers in general; the serial and parallel schedulers probably don't care if they get duplicate batches, they'll just execute the transactions again when they get around to them.

vaporos replied:
It does matter for serial and parallel schedulers, because you wouldn't be able to differentiate the result with the correct one. It might actually be problematic inside the parallel scheduler for other reasons as well, but I think this first reason is sufficient to change the API and make sure we don't allow ingress of duplicate batches. Or more specifically, duplicate batches which are in-progress. (The scheduler shouldn't keep track of batches long-term for this purpose, so it would bascially just ensure all the batches currently known to the scheduler are unique.)

ltseeley replied:
Added a commit that updates add_batch to return a Result. Creates a new SchedulerError enum. Also added a SchedulerFinalized variant that's used instead of panicking when the add_batch method is called on a finalized scheduler.

ltseeley replied:
This update was moved to #85

peterschwarz commented on libtransact/src/scheduler/mod.rs:
```diff
@@ -164,15 +186,15 @@ mod tests {
     pub fn test_scheduler(scheduler: &mut Scheduler) {
         let mut workload = XoBatchWorkload::new_with_seed(5);
-        scheduler.add_batch(workload.next_batch().unwrap());
+        scheduler.add_batch(workload.next_batch().unwrap()).unwrap();
```
For these new unwraps, could you change them to expects, so it's clear what's being tested here.

ltseeley replied:
Done

peterschwarz commented on libtransact/src/scheduler/multi/shared.rs:
```diff
@@ -56,9 +56,15 @@ impl MultiSchedulerShared {
-            scheduler.add_batch(batch.clone());
+            scheduler
+                .add_batch(batch.clone())
+                .expect("Failed to add batch to sub-scheduler");
```
In this case, shouldn't the SchedulerError be propagated?

ltseeley replied:
I debated whether or not to propagate; I decided not to. The multischeduler already checks for duplicate batches, and its check is going to be more strict than any of the sub-schedulers since it will identify a batch as duplicate until all of the sub-schedulers are done with the batch. Furthermore, because the purpose of the multischeduler is to test the sub-schedulers, if the multischeduler doesn't think a batch is duplicate but a sub-scheduler does, the sub-scheduler is almost definitely broken (so it should fail).

vaporos replied:
We might want to keep the multi-scheduler running in LR networks to observe the number of errors - any expects/panics need to be able to be mitigated to accomplish that later.

ltseeley replied:
Fixed to handle errors using the error callback

ltseeley commented:
Updated the PR with the following changes:
- Add a `SubSchedulerHandler` trait to the MultiScheduler's module, which is implemented by the Executor. This removes MultiScheduler -> Executor dependency.
- Update the MultiScheduler's take_task_iterator() and new_notifier() methods to be unimplemented, and indicates that these methods should not be used; instead, the MultiScheduler passes all of the sub-schedulers' task iterators and notifiers to the SubSchedulerHandler.
- Update the MultiScheduler's shutdown method to return a `std::thread::Result<()>` instead of panicking when joining the background thread fails; this allows for making sure the background thread panics in the unit tests.
- Add unit tests that verify the specific behaviors of the MultiScheduler itself.

vaporos commented on libtransact/src/scheduler/multi/core.rs:
```diff
@@ -0,0 +1,131 @@
+/// An enum of messages which can be sent to the MultiSchedulerCore via a
+/// `Sender<MultiSchedulerCoreMessage>`
+pub enum MultiSchedulerCoreMessage {
+    /// The sub-scheduler at the given index in the sub-scheudlers list has gotten the given batch
```
Typo here - 'scheudlers'

ltseeley replied:
Fixed

vaporos commented on libtransact/src/scheduler/multi/core.rs:
```diff
@@ -0,0 +1,131 @@
+                        .shared_lock
+                        .lock()
+                        .expect("scheduler shared lock is poisoned");
+                    let num_schedudlers = shared.schedulers().len();
```
Typo - scheduldlers

ltseeley replied:
Fixed

vaporos commented on libtransact/src/scheduler/multi/core.rs:
```diff
@@ -0,0 +1,131 @@
+                    // Get the pending result for this batch
+                    let (previous_result, already_executed_schedulers) = pending_results
+                        .get_mut(&batch_result.batch)
+                        .expect("Got callback for batch that's not pending");
```
This could be handled via that error callback mechanism as well.

ltseeley replied:
Updated

vaporos commented on libtransact/src/scheduler/multi/core.rs:
```diff
@@ -0,0 +1,131 @@
+                        Some(result) => {
+                            if &batch_result != result {
+                                // Fail fast if the sub-schedulers do not produce the same result
+                                panic!(
```
This should also be handled via the error callback mechansim.

ltseeley replied:
Updated

vaporos commented on libtransact/src/scheduler/multi/mod.rs:
```diff
@@ -0,0 +1,543 @@
+ * -----------------------------------------------------------------------------
+ */
+
+//! A `Scheduler` which runs multiple sub-schedulers.
```
Probably worth elaborating here.

ltseeley replied:
Added more detail

vaporos commented on libtransact/src/scheduler/multi/mod.rs:
```diff
@@ -0,0 +1,543 @@
+/// struct that implements this trait; this replaces the use of the MultiScheduler's own
+/// take_task_iterator and new_notifier methods.
+pub trait SubSchedulerHandler {
+    fn pass_scheduler(
```
This method should have doc comments.

ltseeley replied:
Updated

vaporos commented on libtransact/src/scheduler/multi/mod.rs:
```diff
@@ -0,0 +1,543 @@
+    fn take_task_iterator(&mut self) -> Box<dyn Iterator<Item = ExecutionTask> + Send> {
+        // The MultiScheduler passes its sub-schedulers' task iterators directly to the
+        // SubSchedulerHandler
+        unimplemented!("MultiScheduler's take_task_iterator method should not be called");
```
This method and new_notifier() will have to work in order to pass this scheduler to an executor.

ltseeley replied:
Updated

vaporos commented on libtransact/src/scheduler/multi/mod.rs:
```diff
@@ -0,0 +1,543 @@
+        fn next(&self) {
+            let mut i = 0;
+            for notifier in &self.notifiers {
+                // println!("Getting notifier {}", i);
```
commented-out-code

ltseeley replied:
Removed

vaporos commented on libtransact/src/scheduler/multi/mod.rs:
```diff
@@ -0,0 +1,543 @@
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::scheduler::serial::tests::MockContextLifecycle;
```
I don't think the multi-scheduler should depend upon serial scheduler tests. Maybe pop this MockContextLifecycle up a level to the shceduler::tests module?

ltseeley replied:
Fixed

vaporos commented on libtransact/src/scheduler/serial/mod.rs:
```diff
@@ -150,17 +157,18 @@ impl Scheduler for SerialScheduler {
 
 #[cfg(test)]
-mod tests {
+pub mod tests {
```
Should keep this private.

ltseeley replied:
Fixed

ltseeley commented on libtransact/src/scheduler/mod.rs:
```diff
@@ -104,6 +107,28 @@ pub enum ExecutionTaskCompletionNotification {
+        match *self {
+            SchedulerError::DuplicateBatch(ref batch_pair) => write!(
+                f,
+                "DuplicateBatch: {}",
```
Dealt with in #85

ltseeley commented on libtransact/src/scheduler/mod.rs:
```diff
@@ -104,6 +107,28 @@ pub enum ExecutionTaskCompletionNotification {
+                "DuplicateBatch: {}",
+                batch_pair.batch().header_signature()
+            ),
+            SchedulerError::SchedulerFinalized => write!(f, "SchedulerFinalized"),
```
Dealt with in #85

ltseeley commented:
Made some significant updates:
- Uses the error callback now
- All results are collected before reporting a successful or unsuccessful batch; if all results match for a batch, the multischeduler's callback is used, otherwise all batch results are returned in an error using the error callback.

ltseeley commented:
Updated the MultiScheduler to set the error callbacks of the subschedulers; the MultiScheduler's core thread will receive these errors and report them using its own error callback.

ltseeley commented:
- Rebased this PR on master to incorporate the changes from #86; updated the MultiScheduler to return a None result when all sub-schedulers are done, except if some sub-scheduler(s) did not return results for one of the pending batches (in that case, returns an error)
- Updated the MultiScheduler unit tests to use channels for properly checking results of execution
